components/Search/SearchBar.tsx

```diff
@@ -0,0 +1,184 @@
+"use client";
+import React, { useState } from "react";
+import debounce from "lodash/debounce";
+import "./searchbar.css";
+import Link from "next/link";
+import { useOutsideClick } from "@/utils/hooks";
+
+interface Results {
+  posts: Post[];
+  tags: Tag[];
+  users: User[];
+}
+interface Post {
+  post_id: number;
+  title: string;
+  description: string;
+}
+interface Tag {
+  tag_id: number;
+  tag: string;
+}
+interface User {
+  user_id: number;
+  display_name: string;
+  bio: string;
+  lw_username: string;
+  profile_image_url: string;
+}
+async function searchPosts(query: string) {
+  const response = await fetch("/api/search", {
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json",
+    },
+    body: JSON.stringify({ query: query }),
+  });
+  return response.json();
+}
+
+async function searchTags(query: string) {
+  const response = await fetch("/api/search/tags", {
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json",
+    },
+    body: JSON.stringify({ query: query }),
+  });
+  return response.json();
+}
+async function searchUsers(query: string) {
+  const response = await fetch("/api/search/users", {
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json",
+    },
+    body: JSON.stringify({ query: query }),
+  });
+  return response.json();
+}
+export default function SearchBar() {
+  const [results, setResults] = useState<Results>({
+    posts: [],
+    tags: [],
+    users: [],
+  });
+  const [show, setShow] = useState<boolean>(false);
+  const [loading, setLoading] = useState<boolean>(false);
+  const ref = useOutsideClick(() => {
+    setShow(false);
+  });
+  const handleSearch = async (e: React.ChangeEvent<HTMLInputElement>) => {
+    const query = e.target.value;
+    if (query.length < 1) {
+      setResults({ posts: [], tags: [], users: [] });
+      setShow(false);
+      return;
+    }
+    setLoading(true);
+    const [posts, tags, users] = await Promise.all([
+      searchPosts(query),
+      searchTags(query),
+      searchUsers(query),
+    ]);
+    setResults({ posts, tags, users });
+    setShow(true);
+    setLoading(false);
+  };
+  // const debouncedSearch =  debounce(handleSearch, 300);
+
+  const debouncedResults = React.useMemo(() => {
+    return debounce(handleSearch, 400);
+  }, []);
+
+  React.useEffect(() => {
+    return () => {
+      debouncedResults.cancel();
+    };
+  });
+  return (
```

Comment:
IMO this has too much inlined logic and would be more readable/maintainable if some of the branches of the conditionals were extracted to separate functions